src/Form/Form.php

```diff
@@ -119,17 +122,17 @@ public function data($defaults = false, bool $includeNulls = true): array
 		$data = $this->values;
 
 		foreach ($this->fields as $field) {
-			if ($field->save() === false || $field->unset() === true) {
-				if ($includeNulls === true) {
-					$data[$field->name()] = null;
-				} else {
-					unset($data[$field->name()]);
-				}
+			if ($field->unset() === true) {
+				$data[$field->name()] = null;
 			} else {
-				$data[$field->name()] = $field->data($defaults);
+				$data[$field->name()] = $field->toStoredValue($defaults);
 			}
 		}
 
+		if ($includeNulls === false) {
+			$data = array_filter($data, fn ($value) => $value !== null);
+		}
+
```

Comment:
`Question:` Not sure about this, but I think before with `$field->save() === false` and `$includeNulls === true` the field would end up as `null` in the array. But with the new code. I think it's not anymore.

Reply:
I've reverted that part. I don't think it's necessary to make the PR more complex here than it already is.